### PR TITLE
[Enhancement] Writer error handling

### PR DESCRIPTION
**Pull request by dannyvassallo**
*Issue #, if available:* N/A

*Description of changes:*

Update the writer to better handle errors rather than panic when used as a library.

---
*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: guard-examples/library/src/main.rs [hunk 1]
```diff
@@ -24,7 +24,10 @@ fn main() -> anyhow::Result<()> {
     let mut reader = Reader::new(ReadBuffer::Cursor(Cursor::new(Vec::from(
         payload.as_bytes(),
     ))));
-    let mut writer = Writer::new_with_err(WriteBuffer::Vec(vec![]), WriteBuffer::Vec(vec![]));
+    let mut writer = Writer::new_with_err(WriteBuffer::Vec(vec![]), WriteBuffer::Vec(vec![]))
+        .unwrap_or_else(|err| {
+            panic!("Error: {}", err);
```
Comment (joshfried-aws) [inline guard-examples/library/src/main.rs:29]: we shouldn't be panicking here, we need to return an error

File: guard/src/rules/errors.rs [hunk 2]
```diff
@@ -57,6 +58,10 @@ pub enum InternalError {
     InvalidKeyType(String),
     #[error("internal error {0}")]
     UnresolvedKeyForReporter(String),
+    #[error("{0}")]
+    FromUtf8Error(#[from] Utf8Error),
+    #[error("{0}")]
+    IncompatibleWriteError(String),
```
Comment (joshfried-aws) [inline guard/src/rules/errors.rs:64]: writer*

File: guard/src/utils/writer.rs [hunk 3]
```diff
@@ -18,58 +19,90 @@ impl Default for Writer {
...
+                    Ok(s) => Ok(s),
+                    Err(e) => Err(Error::InternalError(FromUtf8Error(e.utf8_error()))),
+                }
             }
-            _ => unreachable!(),
+            WriteBuffer::Stdout(..) => Err(Error::InternalError(IncompatibleWriteError(
+                "Unable to call err_to_stripped() on a stdout buffer. Stdout is a write-only stream."
+                    .to_string(),
+            ))),
+            WriteBuffer::Stderr(..) => Err(Error::InternalError(IncompatibleWriteError(
```
Comment (joshfried-aws) [inline guard/src/utils/writer.rs:69]: same

File: guard/src/utils/writer.rs [hunk 4]
```diff
@@ -18,58 +19,90 @@ impl Default for Writer {
...
 
-                String::from_utf8(strip_ansi_escapes::strip(data).unwrap())
+                let stripped = strip_ansi_escapes::strip(data.into_bytes());
+                match String::from_utf8(stripped?) {
+                    Ok(s) => Ok(s),
+                    Err(e) => Err(Error::InternalError(FromUtf8Error(e.utf8_error()))),
+                }
             }
-            _ => unreachable!(),
+            WriteBuffer::Stdout(..) => Err(Error::InternalError(IncompatibleWriteError(
```
Comment (joshfried-aws) [inline guard/src/utils/writer.rs:100]: similar comment to previous function about needing a new error to better represent this

File: guard/src/utils/writer.rs [hunk 5]
```diff
@@ -18,58 +19,90 @@ impl Default for Writer {
...
+                "Unable to strip ANSI escapes from stdout buffer.".to_string(),
+            ))),
+            WriteBuffer::Stderr(..) => Err(Error::InternalError(IncompatibleWriteError(
+                "Unable to strip ANSI escapes from stderr buffer.".to_string(),
```
Comment (joshfried-aws) [inline guard/src/utils/writer.rs:104]: same

File: guard/src/utils/writer.rs [hunk 6]
```diff
@@ -93,11 +126,33 @@ pub enum WriteBuffer {
 }
 
 impl WriteBuffer {
-    fn into_string(self) -> Result<String, FromUtf8Error> {
+    fn is_err(&self) -> bool {
+        matches!(self, WriteBuffer::Stderr(_))
+    }
+    fn into_string(self) -> crate::rules::Result<String> {
```
Comment (joshfried-aws) [inline guard/src/utils/writer.rs:132]: same comment as i left about the errors in the 2 previous functions.

File: guard/src/utils/writer.rs [hunk 7]
```diff
@@ -93,11 +126,33 @@ pub enum WriteBuffer {
...
+            WriteBuffer::Stderr(..) => {
+                let incompatible_write_error = IncompatibleWriteError(
+                    "Unable to call into_string() on a stderr buffer. Stderr is a write-only stream."
+                        .to_string(),
+                );
+                let err = Error::InternalError(incompatible_write_error);
+
+                Err(err)
+            }
+            WriteBuffer::Vec(vec) => match String::from_utf8(vec) {
```
Comment (joshfried-aws) [inline guard/src/utils/writer.rs:152]: Nit: we can just use the `?` operator here

File: guard/src/utils/writer.rs [hunk 8]
```diff
@@ -18,58 +19,90 @@ impl Default for Writer {
...
         self.buffer.into_string()
     }
 
-    pub fn stripped(self) -> Result<String, FromUtf8Error> {
+    pub fn stripped(self) -> crate::rules::Result<String> {
         match self.buffer {
-            WriteBuffer::Vec(vec) => String::from_utf8(strip_ansi_escapes::strip(vec).unwrap()),
+            WriteBuffer::Vec(vec) => {
+                let stripped = strip_ansi_escapes::strip(vec);
+                match String::from_utf8(stripped?) {
```
[the remark this answers is not in the document]
Comment (joshfried-aws) [reply guard/src/utils/writer.rs:84]: Nit: We can also dont need to match on the `String::from_utf8` function call, we can just use the `?` operator

**Review by joshfried-aws — CHANGES_REQUESTED** [review]

File: guard-examples/library/src/main.rs (hunk 1 above)
Comment (joshfried-aws) [reply guard-examples/library/src/main.rs:29]: Whoops, misread the example.

File: guard/src/utils/writer.rs [hunk 9]
```diff
@@ -18,58 +21,82 @@ impl Default for Writer {
...
     }
 
-    pub fn stripped(self) -> Result<String, FromUtf8Error> {
+    pub fn stripped(self) -> crate::rules::Result<String> {
         match self.buffer {
-            WriteBuffer::Vec(vec) => String::from_utf8(strip_ansi_escapes::strip(vec).unwrap()),
+            WriteBuffer::Vec(vec) => {
+                let stripped = strip_ansi_escapes::strip(vec)?;
+
+                String::from_utf8(stripped)
```
Comment (joshfried-aws) [inline guard/src/utils/writer.rs:81]: Are you unable to just use the `?` operator here and let thiserror work its magic?

**Review by joshfried-aws — APPROVED** [review]

**Review by minhphuong98 — APPROVED** [review]

**Pull request merged**